File: lib/java/src/main/java/com/workiva/frugal/transport/monitor/BaseFTransportMonitor.java
package com.workiva.frugal.transport.monitor;

/**
 * BaseFTransportMonitor is a default monitor implementation that attempts to reopen a closed transport with exponential
 * backoff behavior and a capped number of retries. Its behavior can be customized by extending this class and
 * overriding desired callbacks.
 */
public class BaseFTransportMonitor implements FTransportMonitor {

    private static final long DEFAULT_MAX_REOPEN_ATTEMPTS = 60;
    private static final long DEFAULT_INITIAL_WAIT = 2 * 1000;
    private static final long DEFAULT_MAX_WAIT = 2 * 2000;

    protected long maxReopenAttempts;
    protected long initialWait;
    protected long maxWait;

    /**
     * Creates a BaseFTransportMonitor with default reconnect options (attempts to reconnect 60 times with 2 seconds
     * between each attempt).
     */
    public BaseFTransportMonitor() {
        this(DEFAULT_MAX_REOPEN_ATTEMPTS, DEFAULT_INITIAL_WAIT, DEFAULT_MAX_WAIT);
    }

    /**
     * Creates a BaseFTransportMonitor with the specified reconnect behavior.
     *
     * @param maxReopenAttempts the max number of reconnect attempts to make.
     * @param initialWait       the number of milliseconds to wait before attempting to reconnect for the first time.
     * @param maxWait           the max number of milliseconds to wait between reconnect attempts (to cap exponential
     *                          backoff).
     */
    public BaseFTransportMonitor(long maxReopenAttempts, long initialWait, long maxWait) {
        this.maxReopenAttempts = maxReopenAttempts;
        this.initialWait = initialWait;
        this.maxWait = maxWait;
    }

    @Override
    public void onClosedCleanly() {
    }

    @Override
    public long onClosedUncleanly(Exception cause) {
        return maxReopenAttempts > 0 ? initialWait : -1;
    }

    @Override
    public long onReopenFailed(long prevAttempts, long prevWait) {
        if (prevAttempts >= maxReopenAttempts) {
            return -1;
        }

        long nextWait = prevWait * 2;
        if (nextWait > maxWait) {
            nextWait = maxWait;
        }
        return nextWait;
    }

    @Override
    public void onReopenSucceeded() {
    }

}

File: lib/src/transport/base_f_transport_monitor.dart
part of frugal;

/// BaseFTransportMonitor is a default monitor implementation that attempts
/// to reopen a closed transport with exponential backoff behavior
/// and a capped number of retries. Its behavior can be customized by extending
/// this class and overriding desired callbacks.
class BaseFTransportMonitor extends FTransportMonitor {
  static const int DEFAULT_MAX_REOPEN_ATTEMPTS = 60;
  static const int DEFAULT_INITIAL_WAIT = 2000;
  static const int DEFAULT_MAX_WAIT = 2000;

  int _maxReopenAttempts;
  int _initialWait;
  int _maxWait;

  StreamController _onConnectController = new StreamController.broadcast();
  StreamController _onDisconnectController = new StreamController.broadcast();

  Stream get onConnect => _onConnectController.stream;
  Stream get onDisconnect => _onDisconnectController.stream;

  bool _isConnected = true;
  bool get isConnected => _isConnected;

  BaseFTransportMonitor(
      {maxReopenAttempts: DEFAULT_MAX_REOPEN_ATTEMPTS,
      initialWait: DEFAULT_INITIAL_WAIT,
      maxWait: DEFAULT_MAX_WAIT}) {
    this._maxReopenAttempts = maxReopenAttempts;
    this._initialWait = initialWait;
    this._maxWait = maxWait;
  }

  @override
  void onClosedCleanly() {
    _isConnected = false;
    _onDisconnectController.add(null);
  }

  @override
  int onClosedUncleanly(cause) {
    _isConnected = false;
    _onDisconnectController.add(cause);

    return _maxReopenAttempts > 0 ? _initialWait : -1;
  }

  @override
  int onReopenFailed(int prevAttempts, int prevWait) {
    if (prevAttempts >= _maxReopenAttempts) {
      return -1;
    }

    return (prevWait * 2).clamp(0, _maxWait);
  }

  @override
  void onReopenSucceeded() {
    _isConnected = true;
    _onConnectController.add(null);
  }
}
